### PR TITLE
WT-7661 Provide stack trace in Evergreen logs when python test dumps core

### PR DESCRIPTION
**Pull request by Siddhartha8899**

### CODE REVIEW

**Comment by Siddhartha8899** [comment]
To test if the script prints the stack traces, I added an assertion in the `__curfile_insert`, and created a patch build. [Here](https://spruce.mongodb.com/version/62289cfd850e617b6c5c3802/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC). 
Note: Currently the script does not support printing stack traces on MAC and Windows.

**Comment by Siddhartha8899** [comment]
At present, the script prints backtraces on macOS and Linux. 
I didn't see a core file on Windows(as a result of a failing Python test). Hence, I suggest limiting the scope of the ticket to obtain the traces in the logs of failing Python test on Ubunutu and macOS should be good enough. Any thoughts @ajmorton?

**Comment by Siddhartha8899** [comment]
Since the core dumps on macOS hosts have quotes around them, e.g. "dump_test_checkpoint.696.core". This makes it harder to load the core in lldb. I have created [BUILD-14820](https://jira.mongodb.org/browse/BUILD-14820) to get rid of the quotes. 

That said, [here](https://spruce.mongodb.com/task/wiredtiger_macos_1014_unit_test_patch_5e2e0e9d66e7c02b5643257871280e63be420ab5_622dab060305b95c5057740d_22_03_13_08_28_10/logs?execution=0) is the patch on macOS

**Comment by ajmorton** [comment]
@Siddhartha8899 Yep I'm happy with this. Linux covers 95% of our use cases and we can spin of a second ticket for the lower prio Windows changes later.

File: test/evergreen/print_python_stack_trace.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    parser.add_argument('-c', '--core_path',
+                        help='directory path to the core dumps',
+                        required=True)
+    parser.add_argument('-l', '--lib_path', help='library path')
+    args = parser.parse_args()
+
+    # Store the path of the core files as a list.
+    core_files = []
+    regex = re.compile(r'.*dump.*python.*', re.IGNORECASE)
+    for root, dirs, files in os.walk(args.core_path):
```
Comment (ajmorton) [inline test/evergreen/print_python_stack_trace.py:114]: Minor: We can replace `dirs` with `_` here to show it's intentionally ignored.

File: test/evergreen/print_python_stack_trace.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    for core_file_path in core_files:
+        print(border_msg(core_file_path), flush=True)
+        if sys.platform.startswith('linux'):
+            dbg = GDBDumper()
+            dbg.dump(args.executable_path, core_file_path, args.lib_path)
+        elif sys.platform.startswith('darwin'):
+            dbg = LLDBDumper()
+            dbg.dump(args.executable_path, core_file_path)
+
```
Comment (ajmorton) [inline test/evergreen/print_python_stack_trace.py:127]: Let's add a comment here so people know Windows is in the works:
```
elif sys.platform.startswith('win32'):
    # FIXME - Windows to be supported in WT-XXXX
    pass
```

**Review by ajmorton — APPROVED** [review]
Very nice!
Two nitpicks but everything LGTM

File: test/evergreen/print_python_stack_trace.py (hunk 1 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: Updated in [9f20287](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/9f20287a1c6c6defe116476933be9e3c948bd697)

File: test/evergreen/print_python_stack_trace.py (hunk 2 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py:130]: Updated in [9f20287](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/9f20287a1c6c6defe116476933be9e3c948bd697)

Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py:130]: Created [WT-8937](https://jira.mongodb.org/browse/WT-8937)

File: test/evergreen/print_python_stack_trace.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:1]: Suggest using `python3` if there's no plan to support python2.

File: test/evergreen/print_python_stack_trace.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import os
+import re
+import subprocess
+import sys
+from shutil import which
+
+
+def border_msg(msg: str):
+    count = len(msg) + 2
+    dash = "-"*count
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:40]: space beside `*`

File: test/evergreen/print_python_stack_trace.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    return "+{dash}+\n| {msg} |\n+{dash}+".format(dash=dash, msg=msg)
+
+
+class LLDBDumper:
+    """LLDBDumper class."""
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:45]: Can we have a more useful docstring here :)

File: test/evergreen/print_python_stack_trace.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self):
+        self.dbg = self.__find_debugger("lldb")
+
+    @staticmethod
+    def __find_debugger(debugger: str):
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:50]: I'm wondering the purpose of the double underscore here in the method name. If it's to mark an internal method as a convention, then a single underscore should be used. Double underscores have the "side effect" of naming mangling, which I guess is not the desired effect here?

File: test/evergreen/print_python_stack_trace.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        """Find the installed debugger."""
+        return which(debugger)
+
+    def dump(self, exe_path: str, core_path: str):
+        """Dump stack trace."""
+        if self.dbg is None:
+            sys.exit("Debugger gdb not found,"
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:57]: gdb --> lldb?

File: test/evergreen/print_python_stack_trace.py (hunk 3 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: Updated in [d6b8174](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/d6b817445834ae25c1f1a883cb5ee7d2d3860373)

File: test/evergreen/print_python_stack_trace.py (hunk 4 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: updated in [d6b8174](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/d6b817445834ae25c1f1a883cb5ee7d2d3860373)

File: test/evergreen/print_python_stack_trace.py (hunk 5 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: Updated in [d6b8174](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/d6b817445834ae25c1f1a883cb5ee7d2d3860373).

File: test/evergreen/print_python_stack_trace.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    parser.add_argument('-c', '--core_path',
+                        help='directory path to the core dumps',
+                        required=True)
+    parser.add_argument('-l', '--lib_path', help='library path')
+    args = parser.parse_args()
+
+    # Store the path of the core files as a list.
+    core_files = []
+    regex = re.compile(r'.*dump.*python.*', re.IGNORECASE)
+    for root, _, files in os.walk(args.core_path):
```
Comment (lukech) [inline test/evergreen/print_python_stack_trace.py:114]: I'm curious why the 2nd tuple (subdirs) is ignored here.

File: test/evergreen/print_python_stack_trace.py (hunk 7 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: Thanks, good find. Updated in [d6b8174](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/d6b817445834ae25c1f1a883cb5ee7d2d3860373)

File: test/evergreen/print_python_stack_trace.py (hunk 6 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: updated in [d6b8174](https://github.com/wiredtiger/wiredtiger/pull/7645/commits/d6b817445834ae25c1f1a883cb5ee7d2d3860373)
Just out of curiosity what are the side effects of name mangling?

Comment (lukech) [reply test/evergreen/print_python_stack_trace.py]: "side effect" may not be the appropriate description here. If you search "name mangling" in [this](https://docs.python.org/3/tutorial/classes.html) link
```Name mangling is helpful for letting subclasses override methods without breaking intraclass method calls.
```
In your case no subclass is used, but in general double underscores are used when the naming mangling feature is needed.

Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py]: makes sense, thanks!

File: test/evergreen/print_python_stack_trace.py (hunk 8 above)
Comment (Siddhartha8899) [reply test/evergreen/print_python_stack_trace.py:114]: As the `subdirs` tuple is not required. Once the file name matches the regex, we simply join the `root` with the file name.

Comment (lukech) [reply test/evergreen/print_python_stack_trace.py:114]: Never mind. I was thinking of the possibility of the regex matched to a subdir, but likely not ('.*dump.*python.*' will be mapped to a file nearly certainly).

**Review by lukech — APPROVED** [review]
LGTM

**Comment by Siddhartha8899** [comment]
I will merge this ticket today, created [WT-8976](https://jira.mongodb.org/browse/WT-8976) to support macOS.

**Pull request merged**